# riceExplorer

This is a Linux/Bash workflow to to identify haplotype reservoirs among a local genomic rice panel (or other organism). The workflow depends on BCFtools, snpEff and Python3 installed to your Linux environement. You are required to download *.gvcf* files from a global database (e.g. rice; IRRI: http://iric.irri.org/resources/3000-genomes-project) representing global genetic and phenotypic diversity of the study organism. You require associated trait data for the files downloaded from the global database for your trait of interest. You then require a local genomic panel of accessions (*.gvcf*) that you want to mine for potentially beneficial SNPs/haplotypes that corellate with extreme phenotypes from the database. The pipeline initially identifies and clusters functional haplotypes of genome-wide gene regions from the global database and then calculates mean phenotypic associations and evaluates whether they are above or below statistical confidence interval criteria. Subsequently, the pipeline searches your local genomic database to recognise if identified markers are found only among selected non-focal varieties (NFVs) versus elite cultivars (ECs). The basic premise is that most agricultural breeding programmes focus on ECs (which over time may may accumulate deleterious alleles) and may come to neglect potential contributions of other accessions held by the research body. It represents a fully automated bioinformatics pipeline that can easily integrate local breeding program sequence data with international database resources, without relying on any (costly, time-consuming) phenotypic experimental procedure. Thus, identified loci may serve as genetic donors to the development panel. **riceExplorer** can automatically conduct a full genome analysis and produces annotated graphical output of chromosomal maps, potential global diversity sources, and summary tables.
