# riceMiner

This is a Linux/Bash workflow to to identify haplotype reservoirs among a local genomic rice panel. The workflow depends on BCFtools, snpEff and Python3 installed to your Linux environement. You are required to download a representative bash of .gvcf files from a global database of rice (i.e. IRRI: http://iric.irri.org/resources/3000-genomes-project) representing global genetic and phenotypic diversity. You require associated trait data for your trait of interest. You also require a local genomic panel of accessions (.gvcf) that you want to mine for the identified potentially beneficial SNPs/haplotypes. The pipeline operates by identifying and clustering functional haplotypes of specific gene regions from the global database and then calculating mean phenotypic associations and evaluating whether they are above or below statistical confidence interval criteria. Subsequently, the pipeline searches the local genomic database to recognise if identified markers are found only among local landrace varieties and not in core development lines. Thus, identified loci may serve as genetic donors to the development panel. This pipeline is fully functional but is still under development.
